File: core/queue.py
"""PrxyClaude · Priority Queue with Concurrency Control"""

from __future__ import annotations

import asyncio
import time
import uuid
from collections.abc import AsyncIterator, Callable, Coroutine
from dataclasses import dataclass, field
from typing import Any

from core.metrics import record_queued


@dataclass(order=True)
class QueueItem:
    priority: int
    created_at: float
    id: str = field(compare=False)
    tier: str = field(compare=False)
    timeout_at: float = field(compare=False)
    future: asyncio.Future = field(compare=False)
    execute: Callable[[], Coroutine[Any, Any, Any]] = field(compare=False)


class RequestQueue:
    def __init__(
        self, max_size: int = 200, timeout_ms: float = 120_000, max_concurrent: int = 10
    ):
        self._queue: asyncio.PriorityQueue[QueueItem] = asyncio.PriorityQueue(
            maxsize=max_size
        )
        self._timeout_ms = timeout_ms
        self._max_concurrent = max_concurrent
        self._semaphore = asyncio.Semaphore(max_concurrent)
        self._active = 0

    @property
    def depth(self) -> int:
        return self._queue.qsize()

    @property
    def active(self) -> int:
        return self._active

    @property
    def max_concurrent(self) -> int:
        return self._max_concurrent

    def stats(self) -> dict:
        return {
            "depth": self.depth,
            "maxSize": self._queue.maxsize,
            "active": self._active,
            "maxConcurrent": self._max_concurrent,
        }

    async def enqueue(
        self, execute: Callable[[], Coroutine[Any, Any, Any]], tier: str
    ) -> Any:
        TIER_PRIORITY = {"opus": 0, "sonnet": 1, "haiku": 2}
        priority = TIER_PRIORITY.get(tier, 1)
        now = time.time() * 1000

        record_queued()

        future = asyncio.get_event_loop().create_future()
        item = QueueItem(
            priority=priority,
            created_at=now,
            id=f"q_{uuid.uuid4().hex[:12]}",
            tier=tier,
            timeout_at=now + self._timeout_ms,
            future=future,
            execute=execute,
        )

        try:
            self._queue.put_nowait(item)
        except asyncio.QueueFull as err:
            raise RuntimeError("Request queue is full") from err

        async with self._semaphore:
            self._active += 1
            try:
                # Check if we timed out while waiting
                if time.time() * 1000 > item.timeout_at:
                    if not future.done():
                        future.set_exception(TimeoutError("Request timed out in queue"))
                    raise TimeoutError("Request timed out in queue")
                result = await execute()
                if not future.done():
                    future.set_result(result)
                return result
            finally:
                self._active -= 1

    async def enqueue_stream(
        self, execute: Callable[[], AsyncIterator[str]], tier: str
    ) -> AsyncIterator[str]:
        """Enqueue an async generator for streaming execution with concurrency control."""
        now = time.time() * 1000

        record_queued()

        # Wait for concurrency slot
        async with self._semaphore:
            self._active += 1
            try:
                timeout_at = now + self._timeout_ms
                if time.time() * 1000 > timeout_at:
                    raise TimeoutError("Request timed out in queue")
                async for chunk in execute():
                    yield chunk
            finally:
                self._active -= 1


_queue: RequestQueue | None = None


def get_queue() -> RequestQueue:
    global _queue
    if _queue is None:
        _queue = RequestQueue()
    return _queue


def queue_stats() -> dict:
    return get_queue().stats()

File: src/queue.ts
// ─────────────────────────────────────────────────────────────────────────────
// PrxyClaude · Priority Request Queue
// ─────────────────────────────────────────────────────────────────────────────

import type { QueuedRequest, ModelTier } from "./types.js";
import { getConfig } from "./config.js";
import { log } from "./logger.js";
import { recordQueued } from "./metrics.js";
import crypto from "crypto";

const queue: QueuedRequest[] = [];
let activeCount = 0;
let draining = false;

const TIER_PRIORITY: Record<ModelTier, number> = {
  opus: 3,
  sonnet: 2,
  haiku: 1,
};

function sortQueue(): void {
  queue.sort((a, b) => {
    if (b.priority !== a.priority) return b.priority - a.priority;
    return a.createdAt - b.createdAt; // FIFO within same priority
  });
}

async function drain(): Promise<void> {
  if (draining) return;
  draining = true;

  const cfg = getConfig().queue;

  while (queue.length > 0 && activeCount < cfg.maxConcurrent) {
    sortQueue();
    const item = queue.shift()!;

    // Check timeout
    if (Date.now() > item.timeoutAt) {
      item.reject(new Error("Request timed out in queue"));
      continue;
    }

    activeCount++;
    (async () => {
      try {
        const result = await item.execute();
        item.resolve(result);
      } catch (err) {
        item.reject(err instanceof Error ? err : new Error(String(err)));
      } finally {
        activeCount--;
        setImmediate(drain);
      }
    })();
  }

  draining = false;
}

export function enqueue<T>(
  execute: () => Promise<T>,
  tier: ModelTier = "sonnet"
): Promise<T> {
  const cfg = getConfig().queue;

  if (queue.length >= cfg.maxSize) {
    return Promise.reject(
      new Error(`Queue full (${cfg.maxSize} requests). Please retry later.`)
    );
  }

  recordQueued();

  return new Promise<T>((resolve, reject) => {
    const item: QueuedRequest = {
      id: crypto.randomUUID(),
      priority: TIER_PRIORITY[tier],
      tier,
      createdAt: Date.now(),
      timeoutAt: Date.now() + cfg.timeoutMs,
      resolve: resolve as (v: unknown) => void,
      reject,
      execute,
    };
    queue.push(item);
    log("debug", `[queue] enqueued ${item.id} tier=${tier} depth=${queue.length}`);
    setImmediate(drain);
  });
}

export function queueStats() {
  const cfg = getConfig().queue;
  return {
    depth: queue.length,
    active: activeCount,
    maxConcurrent: cfg.maxConcurrent,
    maxSize: cfg.maxSize,
  };
}
